File: cpp/src/readdir.cpp
#include "types.hpp"

#include "util.hpp"

#include <algorithm>

namespace filefs {

std::optional<Directory> FileSystem::opendir(const std::string_view path) {
  if (!state_ || !state_->mounted()) {
    return std::nullopt;
  }

  Directory dir;
  std::uint32_t blockindex = 0U;
  int start = 0;
  if (!path.empty() && path.front() == '/') {
    blockindex = 1U;
    start = 1;
    if (!state_->init_pwdtmp("/")) {
      return std::nullopt;
    }
  } else {
    if (state_->tmp.state == 0U) {
      blockindex = state_->pwd_blockindex;
      if (!state_->init_pwdtmp(state_->pwd)) {
        return std::nullopt;
      }
    } else {
      blockindex = state_->tmp.pwd_blockindex;
      if (!state_->init_pwdtmp(state_->tmp.pwd)) {
        return std::nullopt;
      }
    }
  }

  std::array<char, BlockNameMaxSize + 2> s{};
  int slen = 0;
  for (int i = start; i < static_cast<int>(path.size()); ++i) {
    if (path[static_cast<std::size_t>(i)] == '/') {
      if (slen == 0) {
        continue;
      }
      const std::string_view comp{s.data(), static_cast<std::size_t>(slen)};
      const std::uint32_t index = state_->find_path_blockindex(blockindex, comp);
      if (index < 1U) {
        return std::nullopt;
      }
      blockindex = index;
      slen = 0;
      if (!state_->add_to_pwdtmp(path.size(), comp)) {
        return std::nullopt;
      }
      continue;
    }
    s[static_cast<std::size_t>(slen++)] = path[static_cast<std::size_t>(i)];
    if (slen > static_cast<int>(BlockNameMaxSize)) {
      return std::nullopt;
    }
  }
  if (slen > 0) {
    const std::string_view comp{s.data(), static_cast<std::size_t>(slen)};
    const std::uint32_t index = state_->find_path_blockindex(blockindex, comp);
    if (index < 1U) {
      return std::nullopt;
    }
    blockindex = index;
    if (!state_->add_to_pwdtmp(path.size(), comp)) {
      return std::nullopt;
    }
  }

  if (!state_->readblock(blockindex, dir.block_)) {
    return std::nullopt;
  }
  dir.stop_blockindex_ =
      detail::b4_to_u32(std::span<const std::uint8_t, 4>(dir.block_.data() + 12 + 1 + 14 + 4, 4));
  dir.offset_ =
      detail::b2_to_u16(std::span<const std::uint8_t, 2>(dir.block_.data() + 12 + 1 + 14 + 4 + 4, 2));
  dir.blockindex_ = blockindex;
  dir.searchindex_ = 0;
  dir.absolute_path_ = state_->pwd_tmp;
  dir.open_ = true;
  return dir;
}

std::optional<Dirent> FileSystem::readdir(Directory& dir) {
  if (!state_ || !state_->mounted() || !dir.open_) {
    return std::nullopt;
  }

  auto* block = dir.block_.data();
  int k = detail::BlockHead + dir.searchindex_ * 25;
  if (dir.blockindex_ == dir.stop_blockindex_ && k + 1 >= static_cast<int>(dir.offset_)) {
    return std::nullopt;
  }

  for (;;) {
    if (dir.searchindex_ >= detail::BlockItemMaxCount) {
      const std::uint32_t nextindex = detail::b4_to_u32(std::span<const std::uint8_t, 4>(block + 4, 4));
      if (nextindex == 0U || !state_->readblock(nextindex, dir.block_)) {
        return std::nullopt;
      }
      block = dir.block_.data();
      dir.searchindex_ = 0;
      dir.blockindex_ = nextindex;
      k = detail::BlockHead;
      continue;
    }

    const std::uint8_t state = block[k];
    ++k;
    dir.dirent_.type = (state & 0x01U) == 1U ? DirType::File : DirType::Dir;
    std::fill(dir.dirent_.name.begin(), dir.dirent_.name.end(), '\0');
    std::memcpy(dir.dirent_.name.data(), block + k, BlockNameMaxSize);
    k += static_cast<int>(BlockNameMaxSize);
    const std::string_view name = dir.dirent_.name_view();
    dir.dirent_.namelen = name.size();
    if (name == ".") {
      const std::uint32_t dirblockindex = detail::b4_to_u32(std::span<const std::uint8_t, 4>(block + k, 4));
      if (dirblockindex == 1U) {
        dir.dirent_.type = DirType::Root;
      }
    } else if (name == "..") {
      const std::uint32_t dirblockindex = detail::b4_to_u32(std::span<const std::uint8_t, 4>(block + k, 4));
      if (dirblockindex == 0U) {
        dir.dirent_.type = DirType::Root;
      }
    }
    k += 10;
    ++dir.searchindex_;
    return dir.dirent_;
  }
}

void FileSystem::closedir(Directory& d) noexcept { d.open_ = false; }

}  // namespace filefs

File: zig/src/readdir.zig
const std = @import("std");
const types = @import("types.zig");
const util = @import("util.zig");
const block_mod = @import("block.zig");

pub fn opendir(self: anytype, path: []const u8) ?types.OpenDirResult {
    if (self.fp == null) return null;

    const dir = self.allocator.create(types.Dir) catch return null;
    dir.* = .{};

    var blockindex: u32 = undefined;
    var start: usize = undefined;
    if (path.len > 0 and path[0] == '/') {
        blockindex = 1;
        start = 1;
        if (!util.initPwdTmp(self, "/")) {
            self.allocator.destroy(dir);
            return null;
        }
    } else {
        blockindex = util.currentPwdBlockindex(self);
        if (!util.initPwdTmp(self, util.currentPwd(self))) {
            self.allocator.destroy(dir);
            return null;
        }
        start = 0;
    }

    var s: [types.BlockNameMaxSize + 2]u8 = [_]u8{0} ** (types.BlockNameMaxSize + 2);
    var slen: usize = 0;
    var i: usize = start;
    while (i < path.len) : (i += 1) {
        if (path[i] == '/') {
            if (slen == 0) continue;
            const comp = s[0..slen];
            const index = block_mod.findPathBlockindex(self, blockindex, comp);
            if (index < 1) {
                self.allocator.destroy(dir);
                return null;
            }
            blockindex = index;
            slen = 0;
            if (!util.addToPwdTmp(self, comp)) {
                self.allocator.destroy(dir);
                return null;
            }
            continue;
        }
        s[slen] = path[i];
        slen += 1;
        if (slen > types.BlockNameMaxSize) {
            self.allocator.destroy(dir);
            return null;
        }
    }
    if (slen > 0) {
        const index = block_mod.findPathBlockindex(self, blockindex, s[0..slen]);
        if (index < 1) {
            self.allocator.destroy(dir);
            return null;
        }
        blockindex = index;
        if (!util.addToPwdTmp(self, s[0..slen])) {
            self.allocator.destroy(dir);
            return null;
        }
    }

    if (!block_mod.readblock(self, blockindex, dir.block[0..])) {
        self.allocator.destroy(dir);
        return null;
    }
    var b4: [4]u8 = undefined;
    var b2: [2]u8 = undefined;
    @memcpy(b4[0..], dir.block[12 + 1 + 14 + 4 .. 12 + 1 + 14 + 8]);
    dir.stop_blockindex = util.b4ToU32(b4[0..]);
    @memcpy(b2[0..], dir.block[12 + 1 + 14 + 4 + 4 .. 12 + 1 + 14 + 4 + 6]);
    dir.offset = util.b2ToU16(b2[0..]);
    dir.blockindex = blockindex;
    dir.searchindex = 0;
    return .{ .dir = dir, .absolute_path = self.pwd_tmp[0..self.pwd_tmp_len] };
}

pub fn readdir(self: anytype, dir: *types.Dir) ?*types.Dirent {
    if (self.fp == null) return null;

    var block = dir.block[0..];
    var k: usize = types.BlockHead + dir.searchindex * 25;
    if (dir.blockindex == dir.stop_blockindex and k + 1 >= util.u16ToUsize(dir.offset)) {
        return null;
    }
    var b4: [4]u8 = undefined;
    while (true) {
        if (dir.searchindex >= types.BlockItemMaxCount) {
            const nextindex = util.b4ToU32(block[4..8]);
            if (nextindex == 0) return null;
            if (!block_mod.readblock(self, nextindex, dir.block[0..])) return null;
            block = dir.block[0..];
            dir.searchindex = 0;
            dir.blockindex = nextindex;
            k = types.BlockHead;
            continue;
        }

        const state = block[k];
        k += 1;
        const dir_file = state & 0x01;
        dir.dirp.d_type = if (dir_file == 1) types.DTFile else types.DTDir;
        @memset(dir.dirp.d_name[0..], 0);
        @memcpy(dir.dirp.d_name[0..types.BlockNameMaxSize], block[k .. k + types.BlockNameMaxSize]);
        k += types.BlockNameMaxSize;
        const name = util.fixedCStr(dir.dirp.d_name[0..]);
        dir.dirp.d_namlen = name.len;
        if (std.mem.eql(u8, name, ".")) {
            @memcpy(b4[0..], block[k .. k + 4]);
            if (util.b4ToU32(b4[0..]) == 1) dir.dirp.d_type = types.DTRoot;
        } else if (std.mem.eql(u8, name, "..")) {
            @memcpy(b4[0..], block[k .. k + 4]);
            if (util.b4ToU32(b4[0..]) == 0) dir.dirp.d_type = types.DTRoot;
        }
        k += 10;
        dir.searchindex += 1;
        return &dir.dirp;
    }
}

pub fn closedir(self: anytype, dir: *types.Dir) void {
    self.allocator.destroy(dir);
}
